### ISSUE TITLE
Pixel perfect pen option

### ISSUE THREAD

**Issue by flabbet** [PixiEditor/PixiEditor#47]
**Is your feature request related to a problem? Please describe.**
[Pixel perfect](https://deepnight.net/blog/tools/pixel-perfect-drawing/) is an option that is extremely helpful to pixel-art artists. 

**Describe the solution you'd like**
Pixel perfect algorithm that is toggled in Pen Tool Toolbar.

**Comment by MaikelVeen** [comment]
Hi there! For Hacktoberfest I started to look into this issue and ran into some problems. The way this could be implemented is to check for L shapes in an array of pixel coordinates and removing them when this occurs. 

```csharp
private IEnumerable<Coordinates> PixelPerfect(Coordinates[] coordinates)
        {
            int numberOfPixels = coordinates.Length;
            if (numberOfPixels == 1 || numberOfPixels == 0)
            {
                return coordinates;
            }

            List<Coordinates> result = new List<Coordinates>();
            int iterator = 0;

            while (iterator < numberOfPixels)
            {
                if (iterator > 0 && iterator + 1 < numberOfPixels &&
                    (coordinates[iterator - 1].X == coordinates[iterator].X ||
                     coordinates[iterator - 1].Y == coordinates[iterator].Y) &&
                    (coordinates[iterator + 1].X == coordinates[iterator].X ||
                     coordinates[iterator + 1].Y == coordinates[iterator].Y) &&
                    coordinates[iterator - 1].X != coordinates[iterator + 1].X &&
                    coordinates[iterator - 1].Y != coordinates[iterator + 1].Y)
                {
                    iterator += 1;
                }

                result.Add(coordinates[iterator]);
                iterator += 1;
            }

            return result;
        }
```
However it seems  that the PenTool is just a LineTool in disguise. It uses the LineTool to create lines between the current mouse position and the last position:

```
public BitmapPixelChanges Draw(Coordinates startingCoords, Coordinates latestCords, Color color, int toolSize)
        {
            LineTool line = new LineTool();
            return BitmapPixelChanges.FromSingleColoredArray(
                line.CreateLine(startingCoords, latestCords, toolSize), color);
        }
```
For the above solution to work ideally you would want to have all the pixel positions of a line to check and not just the latest. I tried to do this by setting the PenTool to use preview and draw the whole line and checking the pixel perfect afterwards. This works but after a short while the programs starts to lag.

I will be looking into a way tonight to implement it in a performant way, but have not figured out the best solution. Any suggestions?

**Comment by flabbet** [comment]
Hi, sorry for the late response, any progress?

**Comment by flabbet** [comment]
Solved in #149
